RecoEgamma/EgammaIsolationAlgos/src/EleTkIsolFromCands.cc

```diff
@@ -63,38 +62,90 @@ edm::ParameterSetDescription EleTkIsolFromCands::pSetDescript()
 std::pair<int,double> 
 EleTkIsolFromCands::calIsol(const reco::TrackBase& eleTrk,
 			    const pat::PackedCandidateCollection& cands,
-			    const edm::View<reco::GsfElectron>& eles)
+			    const PIDVeto pidVeto)const
 {
-  return calIsol(eleTrk.eta(),eleTrk.phi(),eleTrk.vz(),cands,eles);
+  return calIsol(eleTrk.eta(),eleTrk.phi(),eleTrk.vz(),cands,pidVeto);
 }
 
 std::pair<int,double> 
 EleTkIsolFromCands::calIsol(const double eleEta,const double elePhi,
 			    const double eleVZ,
 			    const pat::PackedCandidateCollection& cands,
-			    const edm::View<reco::GsfElectron>& eles)
+			    const PIDVeto pidVeto)const
 {
-
   double ptSum=0.;
   int nrTrks=0;
 
   const TrkCuts& cuts = std::abs(eleEta)<1.5 ? barrelCuts_ : endcapCuts_;
   
   for(auto& cand  : cands){
-    if(cand.charge()!=0){
-      const reco::Track& trk = cand.pseudoTrack(); 
-      double trkPt = std::abs(cand.pdgId())!=11 ? trk.pt() : getTrkPt(trk,eles);
-      if(passTrkSel(trk,trkPt,cuts,eleEta,elePhi,eleVZ)){	
-	ptSum+=trkPt;
+    if(cand.hasTrackDetails() && cand.charge()!=0 && passPIDVeto(cand.pdgId(),pidVeto)){
+      const reco::Track& trk = cand.pseudoTrack();
+      if(passTrkSel(trk,trk.pt(),cuts,eleEta,elePhi,eleVZ)){	
+	ptSum+=trk.pt();
 	nrTrks++;
       }
     }
   }
   return {nrTrks,ptSum};	
 }
-	
 
-bool EleTkIsolFromCands::passTrkSel(const reco::Track& trk,
+
+
+std::pair<int,double> 
+EleTkIsolFromCands::calIsol(const reco::TrackBase& eleTrk,
+			    const reco::TrackCollection& tracks)const
+{
+  return calIsol(eleTrk.eta(),eleTrk.phi(),eleTrk.vz(),tracks);
+}
+
+std::pair<int,double> 
+EleTkIsolFromCands::calIsol(const double eleEta,const double elePhi,
+			    const double eleVZ,
+			    const reco::TrackCollection& tracks)const
+{
+  double ptSum=0.;
+  int nrTrks=0;
+
+  const TrkCuts& cuts = std::abs(eleEta)<1.5 ? barrelCuts_ : endcapCuts_;
+  
+  for(auto& trk  : tracks){
+    if(passTrkSel(trk,trk.pt(),cuts,eleEta,elePhi,eleVZ)){	
+      ptSum+=trk.pt();
+      nrTrks++;
+    }
+  }
+  return {nrTrks,ptSum};	
+}	
+
+bool EleTkIsolFromCands::passPIDVeto(const int pdgId,const EleTkIsolFromCands::PIDVeto veto)
+{
+  int pidAbs = std::abs(pdgId);
+  switch (veto){
+  case PIDVeto::NONE:
+    return true;
+  case PIDVeto::ELES:
+    if(pidAbs==11) return false;
+    else return true;
+  case PIDVeto::NONELES:
+    if(pidAbs==11) return true;
+    else return false;
+  default:
+    throw cms::Exception("CodeError") <<"invalid PIDVeto "<<static_cast<int>(veto)<<", EleTkIsolFromCands::PIDVeto has been updated without EleTkIsolFromCands::passPIDVeto also being updated";
```

Comment:
is default really needed?
the PIDVeto has no other values.

please add a linebreak

Reply:
It doesnt have any other value currently. This is does not mean that it wont have any other value in the future. In general you will see this approach whenever I do a string to enum conversion (or indeed anything) because I want to ensure that if somebody modifies my code in the future, they also modify this function as well otherwise nasty bugs can slip in.

Reply:
makes sense for string to enum conversions.
Here we have an enum class and all values are known at compie time
It looks like here smth like ``` static_assert(false, "Invalid PIDVeto!");``` is more appropriate because it will be caught at compile time.

Reply:
Ah, I see. This was simply a lack of c++ knowledge on my part. This is better. Changed as you suggested, thanks!

Reply:
okay I couldnt get that to work, I couldnt figure out where to stick the static assert. I did fix things by removing the default and putting the throw outside the switch. 

This now achieves two things:
1) its now an error (well warning promoted to an error) at compile time if this function is not updated when a new enum value is in
2) it protects against very naughty people doing naught static casts from ints.